### PR TITLE
Add ability to save as EPUB as a service

### PR DESCRIPTION
**Pull request by iherman**

### CODE REVIEW

File: js/ui/save-html.js [hunk 1]
```diff
@@ -204,6 +212,24 @@ define(
                 x.document.write(str);
                 x.document.close();
                 x.document.form.submit();
+            }
+            // Create and download an EPUB 3 version of the content
+            // Using (by default) the EPUB 3 conversion service set up at labs.w3.org/epub-generator
+            // For more details on that service, see https://github.com/iherman/respec2epub
+        ,   toEPUB:  function () {
+                respecEvents.pub("save", "toEPUB")
+                var base = window.location.href.replace(/\/[^\/]*$/, "/")
```
Comment (marcoscaceres) [inline js/ui/save-html.js]: This is never used, actually.

**Comment by marcoscaceres** [comment]
@iherman, I rewrote this so it doesn't need to use a different page. The epub service is a GET request, so it doesn't need a form. It just needs the user to click on the link.

**Comment by iherman** [comment]
Thanks @marcoscaceres !

**Comment by marcoscaceres** [comment]
@tobie, could you give this a once over? It's only about 20 lines.

File: js/ui/save-html.js [hunk 2]
```diff
@@ -31,15 +31,19 @@ define(
                     ,   textAlign:      "center"
                     ,   fontSize:       "inherit"
                     }
-                ,   addButton = function (title, content, fileName, popupContent) {
+                ,   addButton = function (title, href, fileName, popupContent, type) {
+                        if(!type){
+                            href = "data:text/html;charset=utf-8," + encodeURIComponent(href)
+                       }
```
Comment (tobie) [inline js/ui/save-html.js]: ws

File: js/ui/save-html.js [hunk 3]
```diff
@@ -31,15 +31,19 @@ define(
                     ,   textAlign:      "center"
                     ,   fontSize:       "inherit"
                     }
-                ,   addButton = function (title, content, fileName, popupContent) {
+                ,   addButton = function (title, href, fileName, popupContent, type) {
```
Comment (tobie) [inline js/ui/save-html.js]: Would suggest refactoring the `addButton` method to take an object with named props at this point.

But feel free to commit as is.

File: js/ui/save-html.js [hunk 4]
```diff
@@ -75,6 +80,17 @@ define(
                 }
                 ui.freshModal("Save Snapshot", $div);
             }
+        // Create and download an EPUB 3 version of the content
+        // Using (by default) the EPUB 3 conversion service set up at labs.w3.org/epub-generator
+        // For more details on that service, see https://github.com/iherman/respec2epub
+        ,   makeEPubHref: function(){
+                if (!conf.epubGenTool) {
+                    conf.epubGenTool = "https://labs.w3.org/epub-generator/cgi-bin/epub-generator.py";
```
Comment (tobie) [inline js/ui/save-html.js]: Is it worth making this an option? Are there other online services that do EPUB generation from HTML?

File: js/ui/save-html.js [hunk 5]
```diff
@@ -75,6 +80,17 @@ define(
...
+                }
+                var epubHref = conf.epubGenTool;
+                epubHref+= "?type=respec&url=" + doc.location.href;
```
Comment (tobie) [inline js/ui/save-html.js]: Don't you want to escape `doc.location.href`, here?

**Comment by tobie** [comment]
Some minor comments, otherwise LGTM.

File: js/ui/save-html.js (hunk 4 above)
Comment (iherman) [inline js/ui/save-html.js]: At the moment, probably not. But there may be more, and it doesn't cost too much to keep that option open...

File: js/ui/save-html.js (hunk 5 above)
Comment (iherman) [inline js/ui/save-html.js]: I am not sure what you mean... What the service needs is a form that can be reused in another GET service right away (the way it works is that it pushes the respec source through the separate spec generator service that Robin set up), and I am not sure that escape would not create problems. That value is just reused by the script right away...

File: js/ui/save-html.js (hunk 4 above)
Comment (tobie) [inline js/ui/save-html.js]: Well, it assumes the API will be the exact same. If you really want to keep your options open, you should be providing a function that calls a callback when it's done with the processing. But then that doesn't serialize well (which might be an issue now that we're serializing the config in the target doc).

File: js/ui/save-html.js (hunk 5 above)
Comment (tobie) [inline js/ui/save-html.js]: What I'm suggesting is simply:

```js
 epubHref+= "?type=respec&url=" + encodeURIComponent(doc.location.href);
```

File: js/ui/save-html.js (hunk 4 above)
Comment (iherman) [inline js/ui/save-html.js]: > On 08 Sep 2015, at 10:35 , Tobie Langel <notifications@github.com> wrote:
> 
> In js/ui/save-html.js <https://github.com/w3c/respec/pull/487#discussion_r38901367>:
> 
> > @@ -75,6 +80,17 @@ define(
> >                  }
> >                  ui.freshModal("Save Snapshot", $div);
> >              }
> > +        // Create and download an EPUB 3 version of the content
> > +        // Using (by default) the EPUB 3 conversion service set up at labs.w3.org/epub-generator
> > +        // For more details on that service, see https://github.com/iherman/respec2epub
> > +        ,   makeEPubHref: function(){
> > +                if (!conf.epubGenTool) {
> > +                    conf.epubGenTool = "https://labs.w3.org/epub-generator/cgi-bin/epub-generator.py";
> Well, it assumes the API will be the exact same. If you really want to keep your options open, you should be providing a function that calls a callback when it's done with the processing. But then that doesn't serialize well (which might be an issue now that we're serializing the config in the target doc).
> 
> 

That is true.

To be honest, I do not really feel too strongly about this…


----
Ivan Herman, W3C
Digital Publishing Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

File: js/ui/save-html.js (hunk 5 above)
Comment (iherman) [inline js/ui/save-html.js]: > On 08 Sep 2015, at 10:38 , Tobie Langel <notifications@github.com> wrote:
> 
> In js/ui/save-html.js <https://github.com/w3c/respec/pull/487#discussion_r38901622>:
> 
> > @@ -75,6 +80,17 @@ define(
> >                  }
> >                  ui.freshModal("Save Snapshot", $div);
> >              }
> > +        // Create and download an EPUB 3 version of the content
> > +        // Using (by default) the EPUB 3 conversion service set up at labs.w3.org/epub-generator
> > +        // For more details on that service, see https://github.com/iherman/respec2epub
> > +        ,   makeEPubHref: function(){
> > +                if (!conf.epubGenTool) {
> > +                    conf.epubGenTool = "https://labs.w3.org/epub-generator/cgi-bin/epub-generator.py";
> > +                }
> > +                var epubHref = conf.epubGenTool;
> > +                epubHref+= "?type=respec&url=" + doc.location.href;
> What I'm suggesting is simply:
> 
>  epubHref+= "?type=respec&url=" + encodeURIComponent(doc.location.href);

Not knowing the deep internals of respec: what exactly does it do? Just the percent encoding stuff? What does that buy me?


----
Ivan Herman, W3C
Digital Publishing Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

File: js/ui/save-html.js (hunk 4 above)
Comment (tobie) [inline js/ui/save-html.js]: Neither do I. Adding that option now just means you'll need to support it in the future, even if by then you realize the API for other services is completely different and you need a function on top of supporting a URL.

File: js/ui/save-html.js [hunk 6]
```diff
@@ -31,15 +31,19 @@ define(
...
                             $("<a></a>")
                                 .appendTo($div)
                                 .text(title)
                                 .css(buttonCSS)
                                 .attr({
-                                    href:   "data:text/html;charset=utf-8," + encodeURIComponent(content)
-                                ,   download:   fileName
+                                    href: href
+                                ,   download: fileName
+                                ,   type: type
```
Comment (tobie) [inline js/ui/save-html.js]: Feels slightly weird to have this undefined type here (unsure what jquery does with it).

File: js/ui/save-html.js (hunk 5 above)
Comment (tobie) [inline js/ui/save-html.js]: That's unrelated to Respec, really. This is just to make sure you URL is valid.

File: js/ui/save-html.js (hunk 6 above)
Comment (iherman) [inline js/ui/save-html.js]: > On 08 Sep 2015, at 10:47 , Tobie Langel <notifications@github.com> wrote:
> 
> In js/ui/save-html.js <https://github.com/w3c/respec/pull/487#discussion_r38902443>:
> 
> >                          if (supportsDownload) {
> >                              $("<a></a>")
> >                                  .appendTo($div)
> >                                  .text(title)
> >                                  .css(buttonCSS)
> >                                  .attr({
> > -                                    href:   "data:text/html;charset=utf-8," + encodeURIComponent(content)
> > -                                ,   download:   fileName
> > +                                    href: href
> > +                                ,   download: fileName
> > +                                ,   type: type
> Feels slightly weird to have this undefined type here (unsure what jquery does with it).
> 
> 

That is not my code… I just added the EPUB access, trying to navigate in the unknown waters of Javascript+respec:-). And this was already there.

Ivan

File: js/ui/save-html.js (hunk 5 above)
Comment (iherman) [inline js/ui/save-html.js]: > On 08 Sep 2015, at 10:49 , Tobie Langel <notifications@github.com> wrote:
> 
> In js/ui/save-html.js <https://github.com/w3c/respec/pull/487#discussion_r38902594>:
> 
> > @@ -75,6 +80,17 @@ define(
> >                  }
> >                  ui.freshModal("Save Snapshot", $div);
> >              }
> > +        // Create and download an EPUB 3 version of the content
> > +        // Using (by default) the EPUB 3 conversion service set up at labs.w3.org/epub-generator
> > +        // For more details on that service, see https://github.com/iherman/respec2epub
> > +        ,   makeEPubHref: function(){
> > +                if (!conf.epubGenTool) {
> > +                    conf.epubGenTool = "https://labs.w3.org/epub-generator/cgi-bin/epub-generator.py";
> > +                }
> > +                var epubHref = conf.epubGenTool;
> > +                epubHref+= "?type=respec&url=" + doc.location.href;
> That's unrelated to Respec, really. This is just to make sure you URL is valid.
> 
> 

But isn't the URL valid already if I get that far? After all, this thing comes in at the end of the respec processing, so to say...

File: js/ui/save-html.js (hunk 6 above)
Comment (tobie) [inline js/ui/save-html.js]: Well, it shows up in the diff, so must have been added by someone. :) I suppose this comes from @marcoscaceres.

File: js/ui/save-html.js (hunk 2 above)
Comment (marcoscaceres) [inline js/ui/save-html.js]: fixed

File: js/ui/save-html.js (hunk 6 above)
Comment (marcoscaceres) [inline js/ui/save-html.js]: defaulted it to ""

File: js/ui/save-html.js (hunk 4 above)
Comment (marcoscaceres) [inline js/ui/save-html.js]: Removed it as an option.

File: js/ui/save-html.js (hunk 5 above)
Comment (marcoscaceres) [inline js/ui/save-html.js]: @tobie, good catch!

**Comment by marcoscaceres** [comment]
@tobie, I did the changes you suggested - including options for `addButton()`. Let me know if you think it's ok.

File: js/ui/save-html.js [hunk 7]
```diff
@@ -60,9 +61,51 @@ define(
 
                     }
                 ;
-                addButton("Save as HTML", self.toString(), "Overview.html", function () { self.toHTMLSource(); });
-                addButton("Save as XHTML5", self.toXML(5), "Overview.xhtml", function () { self.toXHTMLSource(5); });
-                addButton("Save as XHTML 1.0", self.toXML(1), "Overview.xhtml", function () { self.toXHTMLSource(1); });
+                var htmlToDataURL = function(data){
```
Comment (tobie) [inline js/ui/save-html.js]: Why not make `htmlToDataUrl` a method of  `this`?

File: js/ui/save-html.js [hunk 8]
```diff
@@ -60,9 +61,51 @@ define(
...
+                var xhtml1Opts ={
+                    title: "Save as XHTML 1.0",
+                    url: htmlToDataURL(self.toXML(1)),
+                    fileName: "Overview.xhtml",
+                    popupContent: function () { self.toXHTMLSource(1); },
+                };
+                addButton(xhtml1Opts);
+
+                // ePub
+                var ePubOpts = {
```
Comment (tobie) [inline js/ui/save-html.js]: I think this makes the intention a lot clearer. I like it. 

I'd change it to the following but YMMV:

```js
 // ePub
addButton({
    title: "Save as EPUB 3",
    url: this.makeEPubHref(),
    fileName: "spec.epub",
    popupContent: function (){
        window.open(self.makeEPubHref(), "_blank");         
    },
    type: "application/epub+zip"
});
```

**Comment by tobie** [comment]
Ship it!

**Comment by marcoscaceres** [comment]
Released as 3.2.70. 

@iherman, could you please do an announcement on spec-prod@w3.org letting people know of the service... caveat being that people have to save from a real server (e.g., from github pages).

**Comment by iherman** [comment]
Done: https://lists.w3.org/Archives/Public/spec-prod/2015JulSep/0028.html

And thanks again for your help!

**Comment by marcoscaceres** [comment]
Awesome! Thanks for doing the hard work on the server side.

**Pull request merged**